docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,328 @@
+---
+jupyter:
+  jupytext:
+    formats: ipynb,md
+    main_language: python
+    text_representation:
+      extension: .md
+      format_name: markdown
+      format_version: '1.3'
+      jupytext_version: 1.16.4
+  kernelspec:
+    display_name: Python 3
+    name: python3
+---
+
+<!-- #region id="bJ5yuIr-M0x0" -->
+## Mosaic GPU Pipelining
+
+This guide covers software pipelining using the Mosaic GPU backend for Pallas.
+
+For a general overview of the pipelining API in Pallas, we recommend that users first read the [Pallas Pipelining API](https://docs.jax.dev/en/latest/pallas/pipelining.html#pallas-pipelining-api). Pipelining in Pallas is programmed explicitly. For those who are familiar with Triton, this is a significant difference in programming model because in Triton, pipelining is an optimization that is inferred automatically by the compiler.
+
+<!-- #endregion -->
+
+```python id="dGAa3iO5DoRT"
+import jax
+from jax import lax
+from jax import numpy as jnp
+from jax.experimental.pallas import mosaic_gpu as plgpu
+from jax.experimental import pallas as pl
+import numpy as np
+```
+
+<!-- #region id="Pv9j90hVyswo" -->
+
+### Pipelining with Mosaic GPU
+
+
+While the Mosaic GPU backend supports pipelining via `pl.pallas_call`, the recommended approach is to use the `emit_pipeline` function to pipeline over sequential loops, and to use `pl.pallas_call` to partition the problem in parallel over the CUDA grid. `emit_pipeline` follows the same API as `pl.pallas_call` except it exposes a few additional GPU-specific options.
+
+```python
+def emit_pipeline(
+    body: Callable[..., None],
+    *,
+    grid: tuple[int, ...],
+    in_specs: Sequence[pallas_core.BlockSpec] = (),
+    out_specs: Sequence[pallas_core.BlockSpec] = (),
+    max_concurrent_steps: int,
+    delay_release: int,
+) -> Callable:
+```
+
+- `body`, `grid` have the same semantics as in `pl.pallas_call`. The `grid` denotes how many invocations of the `body` function to run. In contrast with a CUDA grid, the pipeline grid is guaranteed to run sequentially.
+- `in_specs` and `out_specs` also work the same as `pl.pallas_call`, except they also accept `plgpu.GPUBlockSpec` instances that can be used specify GPU-specific transforms, such as swizzling. See [memory reference transforms](https://docs.jax.dev/en/latest/pallas/gpu/reference.html#memory-reference-transforms) for more detail on available transformations.
+- `max_concurrent_steps` controls the maximum number of pipeline stages to use. Using additional stages will consume more SMEM to hold temporary buffers, so this option should be used carefully.
```

Comment:
Not sure if "pipeline stages" is quite right here. Should we explain this in terms of the number of copy operations running in concurrently?

CC @apaszke

Reply:
I updated this to just say the number of concurrent memory copies in flight.